### ISSUE TITLE
Can't click on things with various Firefox versions and Linux distros

### ISSUE THREAD

**Issue by eehpcm** [openstreetmap/iD#4547]
On the 16th (the day after 2.5.1 was released) iD stopped working properly for me.

Most functionality works.  Background from Bing displays and different backgrounds can be selected. Objects (points/lines/areas) display, labels display.  Map can be slid.  Map can be zoomed with scrollwheel or right-hand controls.  All the right-hand controls work.

Some functionality doesn't work.  The spinny thing never stops spinning.  I left it for hours and it doesn't stop.  At one point, after around 30 minutes, a strip of red along the bottom complained about being unable to connect to API (or something like that), but I can't reproduce that one and it may have been a temporary connectivity issue.  That's a triviality, but there's worse...

Big problems: I can't select objects to edit them.  The top controls (point/line/area) don't operate so I can't add objects.  It's tantalizingly close to being usable without actually being usable.  I can see everything, I just can't add/modify objects.

I've tried this with Firefox 52.4.0 (64-bit) on CentOS 6.8 and on CentOS 7.3.  I even got hold of a laptop with Windows 10 (ugh) and used Edge (double-ugh) and had the same problem.

For a while I thought it might be bit of javascript being loaded from a different server from the rest of it and I was seeing a routeing problem with my ISP or some auto-blocking/tarpitting on a server taking out a block of IPs I happen to be on.  But then...

Today, fumbling with the touchpad on an unfamiliar laptop, I somehow accidentally triggered the help walkthrough (don't ask me how I did that without noticing, I don't know, I only know I hate touchpads).  Clicked to the last bits of the walkthrough to get out of it (I wish "start editing" really went straight to editing instead of launching a mini-walkthrough, or there was a "I came here by mistake, get me outa here right now" button), and when I finally escaped the walkthrough then iD started working properly (the spinny thing continues spinning, but I can edit/add objects).  I've reproduced this on my main computer, so it's consistent.

Note that this fix doesn't survive logging out/back in.

**Comment by bhousel** [comment]
Wow there's a lot to process in this issue.  Sorry that things are not working for you.  For what it's worth, I conducted a mapathon on the 16th with iD 2.5.1 and about 50 people, all using a mix of browsers, and none of them experienced issues like this.  

The loading spinner will spin as long as there are outstanding requests from the OSM API that have not been fulfilled.  The red api bar will appear when the OSM API server is unavailable or rate limited.

My best guess at the cause of these issues are a mix of a poor network connection, old Firefox and CentOS versions, maybe some broken usb drivers or dead mouse batteries could explain the issues with clicking either not working or clicks going to the wrong things.  

If you have issues like this, iD might not be the best editor for your situation.  An editor like JOSM that works offline or over an intermittent network connection might work better.

Closing here, as I don't really know how to begin debugging this issue, but if you can add some error messages or specific details for me to try, please add here and I'll reopen..

**Comment by eehpcm** [comment]
Everything worked absolutely find until the 16th.  I used iD a lot prior to that.  No problems.

Firefox is the most recent available for CentOS 7.3.  It may not have the latest bling but security fixes are backported and its DOM-handling isn't significantly dated.  In any case, I also have this problem with Windows 10/Edge (and that was after the latest patch Tuesday updates).

Network connection is good enough to watch youtube without stuttering.  Usually at low-res because of data caps, but it works at high-res too.  It's not because of a slow connection.

USB drivers are fine.  The mice don't have batteries.  The clicks work fine to drag the map around.  The clicks work fine to operate the controls at the right of iD.  The clicks work fine with every other app on the computer and every other web page I look at.  The clicks are not able to select iD objects or operate the controls at the top even though they suffice to drag the map and operate the controls at the right.  This is not a mouse/USB problem.  This is definitely a "some part of the javascript isn't doing what it should" problem.

As I said, I found a work-around by bringing up the help walkthrough.  Once I exited that, everything about iD worked as it should.  So it's definitely not a connection/mouse/USB issue, because I've since made several successful edits.  Absolutely fine in all respects.  But it doesn't stay fixed.  Shut down and start back up and I have to go through the walkthrough tutorial again.  I've done that enough times to know that it consistently starts up borked and is consistently fixable with the walkthrough.  100% repeatable across two computers, two OSs, two browsers.

Which sorta implies some flag got prodded by the walkthrough.  Which makes me wonder about a race hazard somewhere.  Especially since you say you had a mapathon on the 16th without problems.  Especially as one other mapper tells me he's had the same problem a couple of times since the 15th, but it went away after he closed then re-opened his browser.  Especially as the computers I have at my disposal are rather old and not blindingly fast.  There's a lot of asynchronous stuff going on in the javascript, so my money is definitely on a race hazard.

My computer is tethered to my mobile for internet, so there can be occasional connectivity interruptions, which explains that API error, now you've told me what causes it.  A one-off and not relevant to the problems I have.  A red herring.  Forget I mentioned it.

I don't have a clue how to begin debugging it either. :)  But if you have any suggestions (that don't involve buying a new computer I can't afford), I'll give them a go.  Maybe if you can find a slower computer to play with...

**Comment by bhousel** [comment]
> I don't have a clue how to begin debugging it either. :) But if you have any suggestions (that don't involve buying a new computer I can't afford), I'll give them a go.

I would definitely appreciate any help you can lend - and it sounds like you are great at troubleshooting. 
 You can try:
- Install iD locally or use the debuggable mirror at http://preview.ideditor.com/master/ 
- set breakpoints in the code and see where the clicks are going / why the API calls in `services/osm.js` are failing

It does sound a lot like #4542 (maybe the same issue?)

**Comment by eehpcm** [comment]
Nope, I'm not great at troubleshooting.  For instance, I forgot to mention that I disabled Adblock Plus and uBlock Origin to no effect (first thing I tried).  I'm just about smart enough to understand when I'm on the wrong side of the Dunning-Kruger effect.

I tried the debuggable mirror.  Predictably, it worked.  Which means we're looking at a Heisenbug and almost certainly a timing problem is involved.

Also, I couldn't see what makes that, as opposed to the non-dev version, debuggable.  Unless the difference is squashed javascript as opposed to unsquashed.  If that's the difference, tracking down a timing bug would be very difficult.  Even more so because the bug isn't showing up in the debuggable mirror.

From what I can see, my symptoms aren't the same as #4542, but in case I misunderstood the other issue I've added a screencap of what I see when I hit the problem.  Both issues may have the same ultimate cause, though.

It's getting late here, and I still have things to do before bed.  So it will probably be tomorrow before I can look into it any further.  And maybe by then you'll have figured it out anyway and saved me the trouble. :)

![id_grab](https://user-images.githubusercontent.com/3224921/33042666-8f2a310a-ce3a-11e7-88d8-256385d189a3.png)

**Comment by eehpcm** [comment]
After some playing and pondering, I can give you some additional info that may narrow it down a little.

This is all about the cursor (aka mouse pointer) and where it behaves differently.  I'm comparing standard 2.5.1 (borked) with 2.5.1 after the workaround of going to the walkthrough tutorial.

When the cursor is over a non-active part of the map, both show the grabby hand, indicating I can drag the map around (and I can do so).

When the cursor is over the controls at the right (layers, map data, zoom, etc.), in both it changes to a tappy finger and a tooltip appears.  The controls work.

When the cursor is over the controls at the top (point, line, area), in both it changes to a tappy finger and a tooltip appears.  The controls do nothing in the borked version but function correctly after the workaround.

Interesting is when I hover over an active part of the map (point, line, boundary of an area).  In the workaround version the pointer changes to a mini-arrow with a square box showing the type of object and the object itself is highlighted with a red outer boundary,   In the borked version, the pointer remains a grabby hand and no highlighting occurs.

Clicking on an object produces different results.  In the workaround version, the result is a flashing red boundary around the object and the object details appear in the left pane.  In the borked version no highlighted boundary (whether flashing or not) and no object details in the left pane.

Clicking **and holding** behaves the same as far as the cursor is concerned.  In both versions the "mini-arrow plus box containing object type indicator" turns into just a mini arrow.

That's about it, as far as I can tell.  But may narrow things down a little.

Further thoughts.  I initially suspected a race hazard and would have put money on the return of a JSON request before a handler had been set up for it.  But that is almost certainly wrong because the debugging version runs a lot slower on my computer but works fine.

You think this is similar to two other issues.  The one complaining about the spinner never stopping probably is.  I'm not so sure about the one where objects aren't displaying.  But there's a way to tell.  Ask them if my workaround fixes their problem. Even if they eventually turn out to have different causes, if the workaround fixes it then at least they can do some mapping.

Workaround is click on help.  Click on "sine-wave leading to X-marks-the-spot-on-the-folded-map" to get the walkthrough.  Leave the walkthrough.  iD is now fixed (until you close the browser window).

**Comment by eehpcm** [comment]
I'm glad you re-opened this, makes it a little easier for me to find. :)  The title change is an improvement in one way but not in another.  I get this exact same behaviour (including the workaround that fixes it) in Edge on Win 10.

I suspect the common factor is hardware age.  People tend to put Linux on older hardware that's no longer capable of running newer versions of Windows, so you're seeing this issue crop up on older hardware which just happens to be running Linux.

My main computer has a 1GHz Athlon 64 X2 Dual Core Processor 4000+. The laptop has Intel(R) Pentium(R) Dual  CPU  T3200  @ 2.00GHz, although /proc/cpuinfo tells me it's actually running at 1GHz.  They're not fast.  They have only two cores.  I doubt they've even heard of threading.  If there are race hazards in your code, this is the type of hardware it's going to show up on.

**Comment by analogdash** [comment]
I'm experiencing the same problem and successfully using the same workaround. Was my first time mapping. Happens in both Firefox 57 and Chromium on Linux Mint XFCE 32bit 18.2. I'm on an Acer Aspire One netbook with an Intel Atom processor and screen resolution of 1024 x 600. Hope this helps.

**Comment by manfredbrandl** [comment]
I also have the problem that the spinny thing never stops spinning. Seen with 2.5.1 (but not with 2.5.0) on iOS 10 Safari (latest version on iPad) and Debian Linux 9.2, 32-bit i386 with Chromium and Firefox.

**Comment by eehpcm** [comment]
A couple more observations that might help narrow down where things are going wrong.

First, where it behaves correctly even when borked.  As stated before, hovering over an object does nothing (no red boundary, no info in the left pane), but I just found an exception.  If I hover over the label (street/road name) on a way, it obligingly hides itself just as it would on the working version.  So mouseover is bubbling up/down/sideways (depending on internals of the DOM) to the code.

Second, I was looking at a question in the forums.  The person had given the URL for the node he was talking about: [(https://www.openstreetmap.org/edit?node=4539676292#map=18/36.72868/-4.67321].  When I clicked on it, _mirabile dictu_ everything worked.

**Comment by osmgit** [comment]
Same problem after latest Firefox Update. I'm running Firefox 57 on Linux Xubuntu.

**Comment by eehpcm** [comment]
> Same problem after latest Firefox Update. I'm running Firefox 57 on Linux Xubuntu.

Three questions.

1) Does the workaround fix it for you?  It's a pain but at least it lets you keep mapping.  It would also show if you're hitting the same problem as me or a different problem that happens to resemble it.

2) Are you able to downgrade Firefox to the version you had previously?  If so, does that fix it?  This is an important question if you're able to give it a try.

3) Is your computer old and/or low on memory?  I think that's a contributing factor, and why so few people are seeing this problem, but I could be wrong.

My experience is that this happened without a Firefox update; it happened after iD went to 2.5.1 on the 15th of the month.  I'm on Firefox 52.4.0 on CentOS 6.8 and on 7.3.  It also happened to me with Edge on Win10.  All computers at my disposal are low on memory and use older 64-bit processors.  It's happened to somebody with Safari/iOS and to somebody else with Chromium/Mint.  I think it's a race hazard introduced in iD 2.5.1, so I'd be surprised if it's also caused by upgrading to the latest Firefox (but life is full of surprises, especially when I try to figure out what's wrong with things).

**Comment by osmgit** [comment]
1. Yes going through the walkthrough works.
2. Instead of downgrading Firefox, I tried it out on my faster laptop running exactly the same version of Firefox and Xubuntu and everything worked ok. I also tested it on Windows 8 and up to date Firefox and everything worked ok too.
3. The problem machine is an old EeePC 1015, Atom N550 1.5ghz Dual Core 2GB ram 64bit.

Hope this information helps!

**Comment by eehpcm** [comment]
@osmgit 

Yes, it all helps to understand what's going on.  Thanks for responding.

Do you know when your Firefox upgrade happened?  You're not the only one to report a problem after a Firefox upgrade, but I think it almost certain it was a coincidence of timing.  Your answers make that close to certain, but a date for the upgrade would make it absolutely certain.

You also confirm it's happening on older machines with little RAM.  Which explains why only a few of us are seeing the problem and what a developer would need in order to be able to reproduce the problem.  I think it possible that even on a higher-spec box it may be possible to provoke the problem by having a lot of other apps running so there's a lot of paging slowing things down (but I could be wrong on that, it may be to do with numbers of cores and/or threading capability).

This is a nasty one to figure out, because the debugging version doesn't have the same problem.  It's a Heisenbug. :(

**Comment by osmgit** [comment]
Firefox Quantum was released on November 14th I believe. So I probably updated sometime around then.

I think this might be related to the iD upgrade rather than the Firefox upgrade. I did some mapping at the start of this month and everything was working fine, then I got back into mapping a couple of days ago and this problem occurred.

**Comment by bhousel** [comment]
Hey!  I just closed #4572, and I'm wondering if that might have also fixed this issue with Firefox on Linux or other platforms possibly susceptible to race conditions.  

If you're experiencing this issue, can you test with the iD here:  http://preview.ideditor.com/master/  and let me know if the problems are still occurring?

**Comment by eehpcm** [comment]
I just tried it and it appears to be working fine.

However, isn't that the debuggable version?  You suggested I try that 2 weeks ago, before you made the changes, and it worked for me back then.  So maybe you've fixed the problem, maybe it's just that the debuggable version changes the timings so it works whether you've fixed the problem or not.

Do you have a non-debuggable version of that?

**Comment by bhousel** [comment]
> However, isn't that the debuggable version? You suggested I try that 2 weeks ago, before you made the changes, and it worked for me back then. So maybe you've fixed the problem, maybe it's just that the debuggable version changes the timings so it works whether you've fixed the problem or not.

Oh, sorry, I missed that above.  If the bug isn't reproducible in http://preview.ideditor.com/master, than I'm out of ideas.

**Comment by eehpcm** [comment]
This bug isn't reproducible in [http://preview.ideditor.com/master](http://preview.ideditor.com/master) but that doesn't mean you haven't fixed this bug, as well as #4572.  It just means that I can't tell if you've fixed it or not.  I won't know either way until you either formally release 2.5.2 or put 2.5.2 somewhere I can get at.

It's probably better to release 2.5.2 anyway.  Even if it doesn't fix this bug it will at least reduce the load on the tile servers.  So there's no point delaying things until you've found out how it affects this bug.

**Comment by maxerickson** [comment]
I've had the eternal spinner a couple times (Arch, Firefox 57). Button reload doesn't help, F5 has fixed it.

**Comment by seahorse41** [comment]
I am experiencing the same problem bhousel described.
My hardware is not That old, running 2.0 GHz i686 32-bit with 2Gb of RAM, but yes a single core. My browser has not changed from when it did work, and was installed in September.

My system is running: Linux i686 Debian 8.5 running PaleMoon 27.3.0SSE and the editing page loads in about 10 seconds. I switched from Firefox because this browser has less bloat and runs faster for nearly everything.

I also tried in Linux i686 Mint 17 (dual-boot) running Firefox 54.0 with same results.

I can only edit in Debian 8.5 running the newest Firefox ESR 52.5 but I have to wait 10 seconds between clicks as something in the javascript makes the cpu run at 100% load, and it no longer caches my clicks like it used to.

When fiddling with the mouse repeating the observed changes in the hand cursor, I stumbled on a workaround without loading help. Try this and see if it works for you, it might provide a clue :)
Click on the [-] box on the right side to zoom out, drag the map to pan in a direction a screen height/width. then click on the center [Press + to edit] box. Now click on a roadway and see if it lets you edit it.

I just followed the link to ID master, and after much zooming and panning to get home, it worked fine for me on the first road I clicked. Maybe I just repeated the workaround I found in doing so, since it didn't start zoomed in with edit already enabled.

Any other questions or ideas I could try for you?

**Comment by eehpcm** [comment]
@seahorse41 

Your workaround also works for me.  In fact, I even simplified it.

Zoom out until you get the "Zoom in to edit" box then click on either of the pluses (the one at the right or the one in the box telling you to zoom in).  Fixed.  No dragging necessary (at least not for me).

Which probably explains the master.  You start zoomed out to the max and have to zoom in to do anything.

Which all ties in with bhousel's thought that maybe closing #4572 would fix this.  That problem was deferred tile loading where it was trying to get an unreasonable amount of tiles.  We can't be certain that's also behind this problem until 2.5.2 is released, but it's looking more and more likely.

I'm not sure about your slowdown, unless there's a lot of paging going on for some reason.  Takes a lot of open browser windows for my system to hit that problem, but then it hits a brick wall and occasionally craps itself.  But then I have 3G of RAM and you have only 2G.

**Comment by seahorse41** [comment]
Okay, sounds like the reported bug is getting handled.

As for the slow issue with newest firefox, should I open a new issue? I suspect something in the system libraries were updated this year causing widespread browser slowdowns, and so would not be a problem specific to this project. OSM Editing with older browsers is still snappy and loads fast.
My memory in use numbers are not even over 1Gb, so I doubt adding RAM would be relevant.
Thanks.

**Comment by eehpcm** [comment]
@seahorse41

I've not experienced a system slowdown, but I'm using an older release of CentOS (no systemd).  The iD editor broke for me when it went from 2.5.0 to 2.5.1 with no updates to anything on the computer in between.

I'm not clear, from what you wrote, if the browser that is experiencing slowness also needs the workaround to let you edit.  If it is, it may be the deferred tile fetches are causing the slowdown in the newer browser but not the older ones.  If it's a general slowdown with the newer browser then it probably isn't an iD issue.

To summarize: your guess is as good as mine. :)

**Comment by seahorse41** [comment]
@eehpcm 
The browser experiencing the slowness, heavy cpu load, is Not affected by this bug, so does not need the workaround. It really is off topic here. Is there any way to debug what the firefox-57 browser is doing when it is spinning it's wheels in the IDeditor? What tiles would it be having trouble with?

I can tell that by disabling css, or js, or animations, that it clears up (speeds up) on other pages. I'm at a loss as to what the specific cause is, or who could help figure it out.

**Comment by eehpcm** [comment]
@seahorse41 

The iD master is supposed to be debuggable.  But I never tried debugging because the issue didn't show up for me (I'd stumbled on your workaround without realizing it).

The tiles it's having trouble with are mentioned in issue  #4572.  If I understand it correctly, it is trying to fetch **all** the tiles.  All 315810 of them.  I didn't see network activity or paging that would indicate Firefox was actually fetching them, but I think it was going slightly bananas thinking about trying to fetch them.  Probably building a list of tiles to fetch or something like that.  With other setups it manages to complete whatever task is stalling on our machines and then realize that it doesn't actually need to prefetch 315810 tiles and can manage with actually fetching fewer.  Something like that.  Go read #4572 for more details.

Probably not a lot either of us can do except wait for 2.5.2 to come out and see if fixing the prefetch bug fixes what we're seeing too.  We have two workarounds.  Given the high zoom I usually work with I find my help tutorial workaround slightly faster than your zoom workaround.

**Comment by SomeoneElseOSM** [comment]
For info, I'm seeing this too.  Things that may be relevant:

* Seemingly slow network connection (but P2 can edit without problems)
* Windows 7
* Firefox 57.02
* No enabled browser addons or plugins
* Spinner going around all the time
* Even when I edit at https://master.apis.dev.openstreetmap.org in an area where there is no data to fetch, the same problem occurs.

**Comment by eehpcm** [comment]
2.6.0 fixes this for me.  Thanks.

**Comment by bhousel** [comment]
> 2.6.0 fixes this for me. Thanks.

Great, I'm happy to hear that 👍
